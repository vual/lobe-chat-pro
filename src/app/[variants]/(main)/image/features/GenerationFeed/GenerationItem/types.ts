import { Generation, GenerationBatch } from '@/types/generation';

export interface GenerationItemProps {
  generation: Generation;
  generationBatch: GenerationBatch;
  prompt: string;
}

export interface ActionButtonsProps {
  onCopySeed?: () => void;
  onDelete: () => void;
  onDownload?: () => void;
  seedTooltip?: string;
  showCopySeed?: boolean;
  showDownload?: boolean;
}

export interface SuccessStateProps {
  aspectRatio: string;
  generation: Generation;
  onCopySeed?: () => void;
  onDelete: () => void;
  onDownload: () => void;
  prompt: string;
  seedTooltip?: string;
}

export interface ErrorStateProps {
  aspectRatio: string;
  generation: Generation;
  onCopyError: () => void;
  onDelete: () => void;
}

export interface LoadingStateProps {
  aspectRatio: string;
  generation: Generation;
  onDelete: () => void;
}
