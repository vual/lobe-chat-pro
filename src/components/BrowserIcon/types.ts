import { CSSProperties } from 'react';

export interface SVGComponent {
  className?: string;
  height?: number | string;
  style?: CSSProperties;
  width?: number | string;
}
