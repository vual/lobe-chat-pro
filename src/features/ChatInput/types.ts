export type { ActionKeys } from './ActionBar/config';
