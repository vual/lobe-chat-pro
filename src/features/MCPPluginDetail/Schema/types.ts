export enum ModeType {
  Docs = 'docs',
  JSON = 'json',
}
