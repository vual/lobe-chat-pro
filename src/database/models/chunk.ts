import { cosineDistance, count, sql } from 'drizzle-orm';
import { and, asc, desc, eq, inArray, isNull } from 'drizzle-orm/expressions';
import { chunk } from 'lodash-es';

import { LobeChatDatabase } from '@/database/type';
import { ChunkMetadata, FileChunk } from '@/types/chunk';

import {
  NewChunkItem,
  NewUnstructuredChunkItem,
  chunks,
  embeddings,
  fileChunks,
  files,
  unstructuredChunks,
} from '../schemas';

export class ChunkModel {
  private userId: string;

  private db: LobeChatDatabase;

  constructor(db: LobeChatDatabase, userId: string) {
    this.userId = userId;
    this.db = db;
  }

  bulkCreate = async (params: NewChunkItem[], fileId: string) => {
    return this.db.transaction(async (trx) => {
      if (params.length === 0) return [];

      const result = await trx.insert(chunks).values(params).returning();

      const fileChunksData = result.map((chunk) => ({
        chunkId: chunk.id,
        fileId,
        userId: this.userId,
      }));

      if (fileChunksData.length > 0) {
        await trx.insert(fileChunks).values(fileChunksData);
      }

      return result;
    });
  };

  bulkCreateUnstructuredChunks = async (params: NewUnstructuredChunkItem[]) => {
    return this.db.insert(unstructuredChunks).values(params);
  };

  delete = async (id: string) => {
    return this.db.delete(chunks).where(and(eq(chunks.id, id), eq(chunks.userId, this.userId)));
  };

  deleteOrphanChunks = async () => {
    const orphanedChunks = await this.db
      .select({ chunkId: chunks.id })
      .from(chunks)
      .leftJoin(fileChunks, eq(chunks.id, fileChunks.chunkId))
      .where(isNull(fileChunks.fileId));

    const ids = orphanedChunks.map((chunk) => chunk.chunkId);
    if (ids.length === 0) return;

    const list = chunk(ids, 500);

    await this.db.transaction(async (trx) => {
      await Promise.all(
        list.map(async (chunkIds) => {
          await trx.delete(chunks).where(inArray(chunks.id, chunkIds));
        }),
      );
    });
  };

  findById = async (id: string) => {
    return this.db.query.chunks.findFirst({
      where: and(eq(chunks.id, id)),
    });
  };

  findByFileId = async (id: string, page = 0) => {
    const data = await this.db
      .select({
        abstract: chunks.abstract,
        createdAt: chunks.createdAt,
        id: chunks.id,
        index: chunks.index,
        metadata: chunks.metadata,
        text: chunks.text,
        type: chunks.type,
        updatedAt: chunks.updatedAt,
      })
      .from(chunks)
      .innerJoin(fileChunks, eq(chunks.id, fileChunks.chunkId))
      .where(and(eq(fileChunks.fileId, id), eq(chunks.userId, this.userId)))
      .limit(20)
      .offset(page * 20)
      .orderBy(asc(chunks.index));

    return data.map((item) => {
      const metadata = item.metadata as ChunkMetadata;

      return { ...item, metadata, pageNumber: metadata?.pageNumber } as FileChunk;
    });
  };

  getChunksTextByFileId = async (id: string): Promise<{ id: string; text: string }[]> => {
    const data = await this.db
      .select()
      .from(chunks)
      .innerJoin(fileChunks, eq(chunks.id, fileChunks.chunkId))
      .where(eq(fileChunks.fileId, id));

    return data
      .map((item) => item.chunks)
      .map((chunk) => ({ id: chunk.id, text: this.mapChunkText(chunk) }))
      .filter((chunk) => chunk.text) as { id: string; text: string }[];
  };

  countByFileIds = async (ids: string[]) => {
    if (ids.length === 0) return [];

    return this.db
      .select({
        count: count(fileChunks.chunkId),
        id: fileChunks.fileId,
      })
      .from(fileChunks)
      .where(inArray(fileChunks.fileId, ids))
      .groupBy(fileChunks.fileId);
  };

  countByFileId = async (ids: string) => {
    const data = await this.db
      .select({
        count: count(fileChunks.chunkId),
        id: fileChunks.fileId,
      })
      .from(fileChunks)
      .where(eq(fileChunks.fileId, ids))
      .groupBy(fileChunks.fileId);

    return data[0]?.count ?? 0;
  };

  semanticSearch = async ({
    embedding,
    fileIds,
  }: {
    embedding: number[];
    fileIds: string[] | undefined;
    query: string;
  }) => {
    const similarity = sql<number>`1 - (${cosineDistance(embeddings.embeddings, embedding)})`;

    const data = await this.db
      .select({
        fileId: fileChunks.fileId,
        fileName: files.name,
        id: chunks.id,
        index: chunks.index,
        metadata: chunks.metadata,
        similarity,
        text: chunks.text,
        type: chunks.type,
      })
      .from(chunks)
      .leftJoin(embeddings, eq(chunks.id, embeddings.chunkId))
      .leftJoin(fileChunks, eq(chunks.id, fileChunks.chunkId))
      .leftJoin(files, eq(fileChunks.fileId, files.id))
      .where(fileIds ? inArray(fileChunks.fileId, fileIds) : undefined)
      .orderBy((t) => desc(t.similarity))
      .limit(30);

    return data.map((item) => ({
      ...item,
      metadata: item.metadata as ChunkMetadata,
    }));
  };

  semanticSearchForChat = async ({
    embedding,
    fileIds,
  }: {
    embedding: number[];
    fileIds: string[] | undefined;
    query: string;
  }) => {
    const similarity = sql<number>`1 - (${cosineDistance(embeddings.embeddings, embedding)})`;

    const hasFiles = fileIds && fileIds.length > 0;

    if (!hasFiles) return [];

    const result = await this.db
      .select({
        fileId: files.id,
        fileName: files.name,
        id: chunks.id,
        index: chunks.index,
        metadata: chunks.metadata,
        similarity,
        text: chunks.text,
        type: chunks.type,
      })
      .from(chunks)
      .leftJoin(embeddings, eq(chunks.id, embeddings.chunkId))
      .leftJoin(fileChunks, eq(chunks.id, fileChunks.chunkId))
      .leftJoin(files, eq(files.id, fileChunks.fileId))
      .where(inArray(fileChunks.fileId, fileIds))
      .orderBy((t) => desc(t.similarity))
      // 先放宽到 15
      .limit(15);

    return result.map((item) => {
      return {
        fileId: item.fileId,
        fileName: item.fileName,
        id: item.id,
        index: item.index,
        similarity: item.similarity,
        text: this.mapChunkText(item),
      };
    });
  };

  private mapChunkText = (chunk: { metadata: any; text: string | null; type: string | null }) => {
    let text = chunk.text;

    if (chunk.type === 'Table') {
      text = `${chunk.text}

content in Table html is below:
${(chunk.metadata as ChunkMetadata).text_as_html}
`;
    }

    return text;
  };
}
