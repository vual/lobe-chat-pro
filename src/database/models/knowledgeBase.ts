import { and, desc, eq, inArray } from 'drizzle-orm/expressions';

import { LobeChatDatabase } from '@/database/type';
import { KnowledgeBaseItem } from '@/types/knowledgeBase';

import { NewKnowledgeBase, knowledgeBaseFiles, knowledgeBases } from '../schemas';

export class KnowledgeBaseModel {
  private userId: string;
  private db: LobeChatDatabase;

  constructor(db: LobeChatDatabase, userId: string) {
    this.userId = userId;
    this.db = db;
  }

  // create

  create = async (params: Omit<NewKnowledgeBase, 'userId'>) => {
    const [result] = await this.db
      .insert(knowledgeBases)
      .values({ ...params, userId: this.userId })
      .returning();

    return result;
  };

  addFilesToKnowledgeBase = async (id: string, fileIds: string[]) => {
    return this.db
      .insert(knowledgeBaseFiles)
      .values(fileIds.map((fileId) => ({ fileId, knowledgeBaseId: id, userId: this.userId })))
      .returning();
  };

  // delete
  delete = async (id: string) => {
    return this.db
      .delete(knowledgeBases)
      .where(and(eq(knowledgeBases.id, id), eq(knowledgeBases.userId, this.userId)));
  };

  deleteAll = async () => {
    return this.db.delete(knowledgeBases).where(eq(knowledgeBases.userId, this.userId));
  };

  removeFilesFromKnowledgeBase = async (knowledgeBaseId: string, ids: string[]) => {
    return this.db.delete(knowledgeBaseFiles).where(
      and(
        eq(knowledgeBaseFiles.knowledgeBaseId, knowledgeBaseId),
        inArray(knowledgeBaseFiles.fileId, ids),
        // eq(knowledgeBaseFiles.userId, this.userId),
      ),
    );
  };
  // query
  query = async () => {
    const data = await this.db
      .select({
        avatar: knowledgeBases.avatar,
        createdAt: knowledgeBases.createdAt,
        description: knowledgeBases.description,
        id: knowledgeBases.id,
        isPublic: knowledgeBases.isPublic,
        name: knowledgeBases.name,
        settings: knowledgeBases.settings,
        type: knowledgeBases.type,
        updatedAt: knowledgeBases.updatedAt,
      })
      .from(knowledgeBases)
      .where(eq(knowledgeBases.userId, this.userId))
      .orderBy(desc(knowledgeBases.updatedAt));

    return data as KnowledgeBaseItem[];
  };

  findById = async (id: string) => {
    return this.db.query.knowledgeBases.findFirst({
      where: and(eq(knowledgeBases.id, id), eq(knowledgeBases.userId, this.userId)),
    });
  };

  // update
  update = async (id: string, value: Partial<KnowledgeBaseItem>) =>
    this.db
      .update(knowledgeBases)
      .set({ ...value, updatedAt: new Date() })
      .where(and(eq(knowledgeBases.id, id), eq(knowledgeBases.userId, this.userId)));

  static findById = async (db: LobeChatDatabase, id: string) =>
    db.query.knowledgeBases.findFirst({
      where: eq(knowledgeBases.id, id),
    });
}
