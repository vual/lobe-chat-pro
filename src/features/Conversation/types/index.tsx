import { type ActionIconGroupEvent } from '@lobehub/ui';
import { type ChatItemProps } from '@lobehub/ui/chat';
import { FC, ReactNode } from 'react';

import { LLMRoleType } from '@/types/llm';
import { ChatMessage } from '@/types/message';

import { type ActionsBarProps } from '../components/ChatItem/ActionsBar';

export type OnActionsClick = (action: ActionIconGroupEvent, message: ChatMessage) => void;
export type OnAvatarsClick = (role: RenderRole) => ChatItemProps['onAvatarClick'];
export type RenderRole = LLMRoleType | 'default' | 'history' | string;
export type RenderMessage = FC<ChatMessage & { editableContent: ReactNode }>;
export type RenderBelowMessage = FC<ChatMessage>;
export type RenderMessageExtra = FC<ChatMessage>;
export type MarkdownCustomRender = (props: {
  displayMode: 'chat' | 'docs';
  dom: ReactNode;
  id: string;
  text: string;
}) => ReactNode;

export type RenderAction = FC<ActionsBarProps & ChatMessage>;

export type RenderItem = FC<{ key: string } & ChatMessage & ListItemProps>;

export interface ListItemProps {
  groupNav?: ChatItemProps['avatarAddon'];

  renderItems?: {
    [role: RenderRole]: RenderItem;
  };

  /**
   * @description 是否显示聊天项的名称
   * @default false
   */
  showTitle?: boolean;
}
