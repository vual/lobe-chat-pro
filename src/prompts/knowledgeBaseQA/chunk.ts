import { ChatSemanticSearchChunk } from '@/types/chunk';

const chunkPrompt = (item: ChatSemanticSearchChunk) =>
  `<chunk fileId="${item.fileId}" fileName="${item.fileName}" similarity="${item.similarity}" ${item.pageNumber ? ` pageNumber="${item.pageNumber}" ` : ''}>${item.text}</chunk>`;

export const chunkPrompts = (fileList: ChatSemanticSearchChunk[]) => {
  if (fileList.length === 0) return '';

  const prompt = `<retrieved_chunks>
<retrieved_chunks_docstring>here are retrived chunks you can refer to:</retrieved_chunks_docstring>
${fileList.map((item) => chunkPrompt(item)).join('\n')}
</retrieved_chunks>`;

  return prompt.trim();
};
