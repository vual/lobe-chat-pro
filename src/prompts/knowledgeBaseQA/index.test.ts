import { describe, expect, it } from 'vitest';

import { ChatSemanticSearchChunk } from '@/types/chunk';
import { KnowledgeItem, KnowledgeType } from '@/types/knowledgeBase';

import { knowledgeBaseQAPrompts } from './index';

describe('knowledgeBaseQAPrompts', () => {
  // Define test data
  const mockChunks: ChatSemanticSearchChunk[] = [
    {
      id: '1',
      fileId: 'file1',
      fileName: 'test.txt',
      text: 'This is a test chunk',
      similarity: 0.8,
      pageNumber: 1,
    },
  ];

  const mockKnowledge: KnowledgeItem[] = [
    {
      id: 'kb1',
      name: 'Test Knowledge',
      type: KnowledgeType.File,
      fileType: 'txt',
      description: 'Test description',
    },
  ];

  const userQuery = 'What is the test about?';
  const rewriteQuery = 'Could you explain the content of the test?';

  it('should return empty string if chunks is empty', () => {
    const result = knowledgeBaseQAPrompts({
      chunks: [],
      knowledge: mockKnowledge,
      userQuery,
    });

    expect(result).toBe('');
  });

  it('should return empty string if chunks is undefined', () => {
    const result = knowledgeBaseQAPrompts({
      knowledge: mockKnowledge,
      userQuery,
    });

    expect(result).toBe('');
  });

  it('should generate prompt with all parameters', () => {
    const result = knowledgeBaseQAPrompts({
      chunks: mockChunks,
      knowledge: mockKnowledge,
      userQuery,
      rewriteQuery,
    });

    // Verify the prompt structure and content
    expect(result).toMatchSnapshot();
  });

  it('should generate prompt without rewriteQuery', () => {
    const result = knowledgeBaseQAPrompts({
      chunks: mockChunks,
      knowledge: mockKnowledge,
      userQuery,
    });

    expect(result).toContain('<raw_query>What is the test about?</raw_query>');
    expect(result).not.toContain('<rewrite_query>');
  });

  it('should generate prompt without knowledge', () => {
    const result = knowledgeBaseQAPrompts({
      chunks: mockChunks,
      userQuery,
    });

    expect(result).toContain(
      'You are also a helpful assistant good answering questions related to',
    );
    expect(result).not.toContain('<knowledge_bases>');
  });

  it('should handle empty knowledge array', () => {
    const result = knowledgeBaseQAPrompts({
      chunks: mockChunks,
      knowledge: [],
      userQuery,
    });

    expect(result).toContain(
      'You are also a helpful assistant good answering questions related to',
    );
    expect(result).not.toContain('<knowledge_bases>');
  });

  it('should properly escape special characters in input', () => {
    const specialChunks: ChatSemanticSearchChunk[] = [
      {
        id: '1',
        fileId: 'file1',
        fileName: 'test&.txt',
        text: 'This is a test with & < > "quotes"',
        similarity: 0.8,
      },
    ];

    const result = knowledgeBaseQAPrompts({
      chunks: specialChunks,
      userQuery: 'Test with & < > "quotes"',
    });

    expect(result).toContain('test&.txt');
    expect(result).toContain('This is a test with & < > "quotes"');
    expect(result).toContain('Test with & < > "quotes"');
  });

  it('should handle multiple knowledge items', () => {
    const multipleKnowledge: KnowledgeItem[] = [
      {
        id: 'kb1',
        name: 'Knowledge 1',
        type: KnowledgeType.File,
      },
      {
        id: 'kb2',
        name: 'Knowledge 2',
        type: KnowledgeType.KnowledgeBase,
      },
    ];

    const result = knowledgeBaseQAPrompts({
      chunks: mockChunks,
      knowledge: multipleKnowledge,
      userQuery,
    });

    expect(result).toContain('Knowledge 1/Knowledge 2');
    expect(result).toContain('<knowledge id="kb1"');
    expect(result).toContain('<knowledge id="kb2"');
  });
});
