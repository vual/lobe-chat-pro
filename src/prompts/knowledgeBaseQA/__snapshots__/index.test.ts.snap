// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`knowledgeBaseQAPrompts > should generate prompt with all parameters 1`] = `
"<knowledge_base_qa_info>
You are also a helpful assistant good answering questions related to Test Knowledge. And you'll be provided with a question and several passages that might be relevant. And currently your task is to provide answer based on the question and passages.
<knowledge_base_anwser_instruction>
- Note that passages might not be relevant to the question, please only use the passages that are relevant.
- if there is no relevant passage, please answer using your knowledge.
- Answer should use the same original language as the question and follow markdown syntax.
</knowledge_base_anwser_instruction>
<knowledge_bases>
<knowledge_bases_docstring>here are the knowledge base scope we retrieve chunks from:</knowledge_bases_docstring>
<knowledge id="kb1" name="Test Knowledge" type="file" fileType="txt" >Test description</knowledge>
</knowledge_bases>
<retrieved_chunks>
<retrieved_chunks_docstring>here are retrived chunks you can refer to:</retrieved_chunks_docstring>
<chunk fileId="file1" fileName="test.txt" similarity="0.8"  pageNumber="1" >This is a test chunk</chunk>
</retrieved_chunks>
<user_query>
<user_query_docstring>to make result better, we may rewrite user's question.If there is a rewrite query, it will be wrapper with \`rewrite_query\` tag.</user_query_docstring>

<raw_query>What is the test about?</raw_query>
<rewrite_query>Could you explain the content of the test?</rewrite_query>
<user_query>
</knowledge_base_qa_info>"
`;
